### PR TITLE
#3378 Shabondi sink(part3): Release data group resource automatically

### PR DESCRIPTION
**Pull request by vitojeng**
required by #3378

### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #issue_number)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

**Comment by chia7712** [comment]
能否調整一下標題讓他更清楚一點

**Comment by vitojeng** [comment]
> 能否調整一下標題讓他更清楚一點

sure

File: ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala [hunk 1]
```diff
@@ -30,69 +30,94 @@ import com.typesafe.scalalogging.Logger
...
+        log.info("create data group: {}", n)
+        val dataGroup = new DataGroup(n, brokerProps, topicNames, pollTimeout)
+        threadPool.submit(dataGroup.producer)
+        dataGroup
+      }
+    )
 
   def size: Int = dataGroups.size()
 
+  def freeIdleGroup(idleTime: JDuration): Unit = {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: 只是一個疑問: 當要被關閉的datagroup正好要被另一個thread (request)使用的話會發生什麼事情?

Comment (vitojeng) [reply ohara-shabondi/src/main/scala/com/island/ohara/shabondi/sink/SinkDataGroups.scala]: data group 在關閉過程動作就是 stop QueueProducer, 包含關閉內部的 consumer.
這個過程中, 若另一個 thread 正在 poll 這個 queue 的話, 應該會是 queue 有資料仍然可以被取出, 直到被讀完為止.

**Comment by vitojeng** [comment]
retry

**Comment by jackyoh** [comment]
retry

**Comment by vitojeng** [comment]
not related error

**Pull request merged**